libraries/common/array.effekt

```diff
@@ -1,90 +1,114 @@
 module array
 
 import effekt
-import option
+import exception
 import list
 
-/**
- * A mutable 0-based indexed fixed-sized array.
- */
-extern type Array[T]
+// Primitive operations:
 
 /**
- * Creates a new Array of size `size` filled with the value `default`
+ * A mutable 0-indexed fixed-sized array.
  */
-extern io def fill[T](size: Int, default: T): Array[T] =
-  ml "Array.array (${size}, ${default})"
-  default {
-    val arr = emptyArray[T](size)
-    each(0, size) { i =>
-      put(arr, i, default)
-    }
-    return arr;
-  }
-
-/**
- * Builds a new Array of size `size` from a computation `index` which gets an index
- * and returns a value that will be on that position in the resulting array
- */
-def build[T](size: Int) { index: Int => T }: Array[T] = {
-  val arr = emptyArray[T](size)
-  each(0, size) { i =>
-    put(arr, i, index(i))
-  }
-  return arr
-}
+extern type Array[T]
 
 /**
- * Converts a List `list` to an Array
+ * Violated Postcondition: Fields contain values
  */
-def fromList[T](list: List[T]): Array[T] = {
-  val listSize = list.size();
-  val arr = emptyArray(listSize);
-
-  foreachIndex(list) { (i, head) =>
-    put(arr, i, head)
-  }
-  return arr;
-}
-
-extern pure def emptyArray[T](size: Int): Array[T] =
+extern pure def unsafeNew[T](size: Int): Array[T] =
   js "(new Array(${size}))"
   chez "(make-vector ${size})" // creates an array filled with 0s on CS
-
-def get[T](arr: Array[T], index: Int): Option[T] =
-  if (index < 0 || index >= arr.size) None()
-  else undefinedToOption(arr.unsafeGet(index))
+  llvm """
+    %z = call %Pos @c_array_new(%Int %size)
+    ret %Pos %z
+  """
 
 extern pure def size[T](arr: Array[T]): Int =
   js "${arr}.length"
   chez "(vector-length ${arr})"
   ml "Array.length ${arr}"
+  llvm """
+    %z = call %Int @c_array_size(%Pos %arr)
+    ret %Int %z
+  """
 
+/**
+ * Unchecked Precondition: Index in bounds
```

Comment:
Depending on what the construction asserts there is more than this here... (see above)